File: src/main/java/net/yakclient/mixin/internal/loader/ContextPool.java
package net.yakclient.mixin.internal.loader;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.Map;

public abstract class ContextPool {
    private final Map<PackageTarget, Context> targets;

    private final ClassLoader loader;

    public ContextPool() {
        this.targets = new HashMap<>();
        this.loader = ClassLoader.getSystemClassLoader();
    }

    private boolean encapsulatedTargetExists(PackageTarget target) {
        for (PackageTarget packageTarget : this.targets.keySet()) {
            if (packageTarget.isTargetOf(target)) return true;
        }
        return false;
    }

    public PackageTarget encapsulatedTarget(PackageTarget target) {
        if (!this.encapsulatedTargetExists(target)) throw new IllegalStateException("Target MUST be encapsulated!");
        for (PackageTarget packageTarget : this.targets.keySet()) {
            if (packageTarget.isTargetOf(target)) return packageTarget;
        }
        throw new IllegalStateException("Target MUST be encapsulated!");
    }

    public Context addTarget(PackageTarget target, ClassLoader loader) {
        final PackageTarget realTarget = this.getTarget(target);
        return this.targets.put(realTarget, this.createContext(loader, realTarget));
    }

    public Context addTarget(PackageTarget target) {
        final PackageTarget realTarget = this.getTarget(target);
        final ClassLoader loader = this.targets.containsKey(realTarget) ? this.targets.get(realTarget).getLoader() : ContextPoolManager.createLoader(realTarget);
        return this.targets.putIfAbsent(this.getTarget(target), this.createContext(loader, target));
    }

    @Contract(pure = true)
    public PackageTarget getTarget(PackageTarget target) {
        return this.encapsulatedTargetExists(target) ? this.encapsulatedTarget(target) : target;
    }

    public boolean isTargeted(String target) {
        final PackageTarget packageTarget = this.createTarget(target);
        return this.isTargeted(packageTarget);
    }

    public boolean isTargeted(PackageTarget target) {
        for (final PackageTarget pT : this.targets.keySet()) {
            if (pT.isTargetOf(target)) return true;
        }
        return false;
    }

    @Nullable
    public Class<?> loadClassOrNull(String name) {
        final PackageTarget target = this.getTarget(this.createTarget(name));
        return this.targets.get(target).findClass(name);
    }

    private Context createContext(ClassLoader loader, PackageTarget target) {
        return new Context(loader, target, this);
    }

    PackageTarget createTarget(String path) {
        try {
            if (this.loader.loadClass(path) != null) return ClassTarget.create(path);
            return PackageTarget.create(path);
        } catch (ClassNotFoundException e) {
            return PackageTarget.create(path);
        }
    }

    public Class<?> defineClass(ClassTarget target, byte[] bytes) {
        final PackageTarget finalTarget = this.getTarget(target);
        final Context context = this.targets.get(finalTarget);
        final String name = target.toString();

        if (!context.getLoader().isDefined(name)) return context.getLoader().defineClass(name, bytes);
        else this.targets.put(finalTarget, this.createContext(ContextPoolManager.createLoader(finalTarget), finalTarget));

        return this.targets.get(finalTarget).getLoader().defineClass(name, bytes);
    }
}

File: base/src/main/kotlin/net/yakclient/mixins/base/internal/loader/ContextPool.kt
package net.yakclient.mixins.base.internal.loader

import net.yakclient.mixins.base.target.JarTarget
import net.yakclient.mixins.base.target.ModuleTarget
import net.yakclient.mixins.base.target.Target
import java.lang.module.FindException

const val JAVA_TARGET_PROPS = "yakclient.mixins.targets"
const val JAVA_PROPS_DELIMITER = ","

private fun <K, V> Map<K, V>.checkPresent(key: K, message: String): V {
    return checkNotNull(get(key)) { message }
}

abstract class ContextPool<T: Target> {
    protected val targets: MutableMap<T, Context> = HashMap()

    fun targets(): List<T> {
        return targets.keys.toList()
    }

    abstract fun addTarget(target: T)

    protected abstract fun defaultLoader(target: T): ProxyClassLoader

    fun isTargeted(target: T): Boolean {
        return targets.containsKey(target)
    }

    fun defineClass(target: T, name: String, bytes: ByteArray): Class<*> {
        return targets.checkPresent(target, "Failed to find appropriate context to define class in")
            .let {
                if (it.isDefined(name)) Context(defaultLoader(target)).also { context -> targets[target] = context }
                else it
            }.defineClass(name, bytes)
    }

    fun loadClassOrNull(target: T, name: String): Class<*>? {
        return targets.checkPresent(target, "Failed to find context from given target: ${target.name()}")
            .findClass(name)
    }
}

class ModuleContextPool : ContextPool<ModuleTarget>() {
    init {
        System.getProperty(JAVA_TARGET_PROPS).split(JAVA_PROPS_DELIMITER).map { name ->
            ModuleLayer.boot().findModule(name)
                .orElseThrow { FindException("Failed to fine module: $name") }
                .let { module ->
                    ModuleTarget(module!!).also {
                        targets[it] = Context(ModuleTargetLoader(it))
                    }
                }
        }
    }

    override fun addTarget(target: ModuleTarget) {
        throw IllegalAccessError("Targets must have been added via command arguments on startup.")
    }

    override fun defaultLoader(target: ModuleTarget): ProxyClassLoader {
        return ModuleTargetLoader(target)
    }
}

class LibContextPool: ContextPool<JarTarget>() {
    override fun addTarget(target: JarTarget) {
        targets[target] = Context(ExternalJarLoader(target.url))
    }

    override fun defaultLoader(target: JarTarget): ProxyClassLoader {
        return ExternalJarLoader(target.url)
    }
}
//
//abstract class `1ModuleContextPool` {
//    private val targets: MutableMap<Target, Context> =
//        mutableMapOf(*System.getProperty(JAVA_TARGET_PROPS).split(JAVA_PROPS_DELIMITER).map { name ->
//            ModuleLayer.boot().findModule(name)
//                .orElseThrow { FindException("Failed to fine module: $name") }
//                .let { module ->
//                    val target = ModuleTarget(module!!)
//                    Pair(target, Context(ModuleTargetLoader(target = target)))
//                }
//        }.toTypedArray())
//
////    private fun findModule(name: String): ResolvedModule? {
////        for (target in targets) {
////            if (LoaderUtils.packages(target.key).contains(LoaderUtils asClassPackage name)) return target.key
////        }
////        return null
////    }
////
////    private fun findContext(name: String): Context? {
////        for (target in targets) {
////            if (LoaderUtils.packages(target.key).contains(LoaderUtils asClassPackage name)) return target.value
////        }
////        return null
////    }
//
//    fun targets(): List<Target> {
//        return targets.keys.toList()
//    }
//
//    fun loadClassOrNull(target: Target, name: String): Class<*>? {
//        return targets.checkPresent(target, "Failed to find context from given target: ${target.name()}")
//            .findClass(name)
//    }
//
//    fun isTargeted(target: Target): Boolean {
//        return targets.containsKey(target)
//    }
//
//    fun isClassDefined(target: Target, name: String): Boolean {
//        return targets.checkPresent(target, "Failed to find context from given target: ${target.name()}")
//            .isDefined(name)
//    }
//
//    fun defineClass(target: Target, name: String, bytes: ByteArray): Class<*> {
//        return targets.checkPresent(target, "Failed to find appropriate context to define class in")
//            .let {
//                if (it.isDefined(name)) this.setContext(
//                    target,
//                    Context(target.createLoader())
//                ) else it
//            }.defineClass(name, bytes)
//    }
//
//    private fun setContext(target: Target, context: Context): Context {
//        targets[target] = context
//        return context
//    }
//}